lang/cs/Org.Apache.REEF.IO/FileSystem/AzureDataLake/AzureDataLakeFileSystem.cs

```diff
@@ -0,0 +1,230 @@
+        }
+
+        /// <summary>
+        /// Creates a new file under the given URI.
+        /// </summary>
+        /// <exception cref="AdlsException">If the URI couldn't be created.</exception>
+        public Stream Create(Uri fileUri)
+        {
+            return _adlsClient.CreateFile(fileUri.AbsolutePath, IfExists.Overwrite);
+        }
+
+        /// <summary>
+        /// Deletes the file under the given URI.
+        /// </summary>
+        /// <exception cref="IOException">If the specified file cannot be deleted</exception>
+        public void Delete(Uri fileUri)
+        {
+            bool deleteStatus = _adlsClient.Delete(fileUri.AbsolutePath);
+            if (!deleteStatus)
+            {
+                throw new IOException($"Cannot delete directory/file specified by {fileUri}");
+            }
+        }
+
+        /// <summary>
+        /// Determines whether a file exists under the given URI.
+        /// </summary>
+        public bool Exists(Uri fileUri)
+        {
+            return _adlsClient.CheckExists(fileUri.AbsolutePath);
+        }
+
+        /// <summary>
+        /// Copies the file referenced by sourceUri to destinationUri.
+        /// Note : This method reads from the input stream of sourceUri locally and
+        /// writes to the output stream of destinationUri.
+        /// This is time consuming and not recommended for large file transfers.
+        /// </summary>
+        /// <exception cref="IOException">If copy process encounters any exceptions</exception>
+        public void Copy(Uri sourceUri, Uri destinationUri)
+        {
+            try
+            {
+                using (var readStream = Open(sourceUri))
+                {
+                    readStream.Position = 0;
+                    using (var writeStream = Create(destinationUri))
+                    {
+                        readStream.CopyTo(writeStream);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new IOException($"Error copying {sourceUri} to {destinationUri}", ex);
+            }
+        }
+
+        /// <summary>
+        /// Copies the remote file to a local file.
+        /// </summary>
+        /// <exception cref="IOException">If copy process encounters any exceptions</exception>
+        public void CopyToLocal(Uri remoteFileUri, string localName)
+        {
+            try
+            {
+                TransferStatus status = _adlsClient.BulkDownload(remoteFileUri.AbsolutePath, localName); // throws KeyNotFoundException
+                if (status.EntriesFailed.Count != 0)
+                {
+                    throw new IOException($"{status.EntriesFailed.Count} entries did not get transferred correctly");
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new IOException($"Error copying {remoteFileUri} to {localName}", ex);
+            }
```

Comment:
I would avoid wrapping one `IOException` into another, e.g.
```C#
TransferStatus status;
try
{
    status = _adlsClient.BulkDownload(remoteFileUri.AbsolutePath, localName); // throws KeyNotFoundException
}
catch (Exception ex)
{
    throw new IOException($"Error in bulk download from {remoteFileUri} to {localName}", ex);
}
if (status.EntriesFailed.Count != 0)
{
    throw new IOException($"{status.EntriesFailed.Count} entries did not get transferred correctly");
}
```
(same applies to `CopyFromLocal()`)

Reply:
fixed both